### PR TITLE
Load songs asynchronously

### PR DESCRIPTION
**Pull request by m-m-adams**
Load songs asynchronously to allow removing the playback handler call from the load function

Splits song loading into 4 pieces, each of which have a callback to continue when ready rather than busy waiting

### CODE REVIEW

File: src/deluge/gui/ui/load/load_song_ui.cpp [hunk 1]
```diff
@@ -226,233 +227,258 @@ ActionResult LoadSongUI::buttonAction(deluge::hid::Button b, bool on, bool inCar
 	return ActionResult::DEALT_WITH;
 }
 
+enum class LoadStatus { start, loadingSamples, waitingToEnd, newSongPlaying, finished };
```
Comment (sapphire-arches) [inline src/deluge/gui/ui/load/load_song_ui.cpp:230]:
```suggestion
enum class LoadStatus { START, BEGIN_LOADING_SAMPLES, WAIT_FOR_SAMPLES, NEW_SONG_PLAYING, COMPLETE };
```

File: src/deluge/gui/ui/load/load_song_ui.cpp [hunk 2]
```diff
@@ -226,233 +227,258 @@ ActionResult LoadSongUI::buttonAction(deluge::hid::Button b, bool on, bool inCar
...
+StorageManager* bdsm{};
```
Comment (sapphire-arches) [inline src/deluge/gui/ui/load/load_song_ui.cpp:231]: These globals make me feel like we should teach the task manager how to use a `std::function` instead, but it's probably fine for an initial pass

File: src/deluge/gui/ui/load/load_song_ui.cpp [hunk 3]
```diff
@@ -226,233 +227,258 @@ ActionResult LoadSongUI::buttonAction(deluge::hid::Button b, bool on, bool inCar
...
+void LoadSongUI::performLoad(StorageManager& _bdsm) {
+	bdsm = &_bdsm;
+	performLoadFixedSM();
+}
+void LoadSongUI::performLoadFixedSM() {
+	static int32_t count = 0;
+	Song* toDelete = nullptr;
+	static LoadStatus status = LoadStatus::start;
+	if (status == LoadStatus::finished)
+		status = LoadStatus::start;
```
Comment (sapphire-arches) [inline src/deluge/gui/ui/load/load_song_ui.cpp:242]:
```suggestion
	if (status == LoadStatus::finished) {
		status = LoadStatus::start;
	}
```

Style nits aside, should this reset happen in `performLoad` instead with a freeze if we try to start a load while another is ongoing?

File: src/deluge/gui/ui/load/load_song_ui.cpp [hunk 4]
```diff
@@ -226,233 +227,258 @@ ActionResult LoadSongUI::buttonAction(deluge::hid::Button b, bool on, bool inCar
...
 	}
-
-	else {
-		playbackHandler.doSongSwap();
+	case LoadStatus::waitingToEnd: {
+		if (currentUIMode != UI_MODE_LOADING_SONG_NEW_SONG_PLAYING) {
+			addOnceTask([]() { loadSongUI.performLoadFixedSM(); }, 100, 0.05);
+			return;
+		}
+		status = LoadStatus::newSongPlaying;
```
Comment (sapphire-arches) [inline src/deluge/gui/ui/load/load_song_ui.cpp:446]: do we need to add a task here? If not there should be a `[[fallthrough]];`

File: src/deluge/gui/ui/load/load_song_ui.cpp (hunk 2 above)
Comment (m-m-adams) [reply src/deluge/gui/ui/load/load_song_ui.cpp:231]: std::function is huge to accommodate arbitrary sized captures and the type erasure indirection makes it significantly slower than function pointers. We could explore an alternate implementation with limited capture size like this one - https://www.codeproject.com/Articles/7150/Member-Function-Pointers-and-the-Fastest-Possible

**Pull request merged**